keps/sig-api-machinery/2161-apiserver-default-labels/README.md

```diff
@@ -0,0 +1,452 @@
+
+This was initially envisioned as a workaround to a limitation in the network policy API, but has emerged as a potential to other problems which may be solved by default namespace labels.
+
+## Motivation
+
+In https://github.com/kubernetes/enhancements/issues/2112 , we discussed several 
+ways to implement network policies that select namespaces by name, rather than by 
+relying on labels which may be unknown, or uneditable by people writing policies, 
+or untrustworthy.
+
+This was inspired by a broad community ask to target namespaces in this manner, 
+as evidenced by issues such as [kubernetes #88253](https://github.com/kubernetes/kubernetes/issues/88253), 
+and the larger set of user feedback obtained in  [user stories](https://github.com/jayunit100/network-policy-subproject/blob/master/p0_user_stories.md).
+
+It’s also worth noting that there are at least 2 other NamespaceSelector APIs in 
+k/k - Validating and Mutating webhooks.  They will inevitably need similar API 
+evolution, but we don't need to do that as part of this KEP.
+
+It turns out that alternate [solutions](#alternatives):
+Break the old API in fundamental ways that are questionable in terms of cost/benefit, 
+add nesting complexity to the network policy API, wherein certain “subsets” of
+the API were not usable together.  This is not “bad”, but it is suboptimal and
+introduces redundant semantics. This is avoidable if we could assume, in all
+cases, that namespaces have a default label, thereby leveraging existing
+functionalities offered by label selectors.
+
+There may also be many other usecases where the security or permissions boundary
+might be conveniently defined in terms of a namepsace name, and that such a
+definition would be most easily referenced by a default label.
+
+### Goals
+
+- Add the ability to select namespaces by name reliably using traditional label selector methods.
+
+### Non-Goals
+
+* Publishing name labels on all resources, this is an option we can explore later for a larger subset of fields in the Kubernetes API.
+* Exposing arbitrary fields to selectors.
+
+## Proposal
+
+We propose to label all namespaces, by default, with a reserved label of
+“kubernetes.io/metadata.name” and set it to the namespace's name. This will allow
+any resource with a `namespaceSelector` field (such as NetworkPolicy, MutatingWebhookConfiguration
+etc) to select (or exclude with `notIn`) these namespaces by explicitly matching on
+the namespace's name with traditional label selector mechanisms.
+
+- If this label is missing, it is added by the apiserver as a default.
+- If this label is deleted, it is added by the apiserver as a default.
+- defaults.go mutates namespaces on read, to always have the label `kubernetes.io/metadata.name=obj.Name`.
```

Comment:
Using a defaulter, if someone tries to update it, would they get a success return code and then have their change not apply?  If so, could you clarify that behavior here.

Reply:
yeah - you would never really be able to mutate the label.

Reply:
ill updated the KEP to clarify this